File: Problems/No.649_Dota2_Senate/dota2_senate.py
class Solution(object):
    def predictPartyVictory(self, senate):
        """
        :type senate: str
        :rtype: str
        """

        # parameters
        Radiants, Dires = [], []
        senate_size = len(senate)

        # start computing
        for i, c in enumerate(senate):
            if (c == 'R'):
                Radiants.append(i)
            else:
                Dires.append(i)

        while (Radiants and Dires):
            curr_R = Radiants.pop(0)
            curr_D = Dires.pop(0)

            if (curr_R < curr_D):
                Radiants.append(senate_size + curr_R)
            else:
                Dires.append(senate_size + curr_D)

        return "Radiant" if Radiants else "Dire"

File: Problems/No.649_Dota2_Senate/dota2_senate.cpp
class Solution {
public:
    string predictPartyVictory(string senate) {
        
        // parameters
        std::queue<int> Radiants, Dires;
        int senate_size = senate.length();

        // start computing
        for (int i=0; i<senate_size; i++) {

            if (senate[i] == 'R') Radiants.push(i);
            else Dires.push(i);

        }
        while (!Radiants.empty() && !Dires.empty()) {

            int curr_R = Radiants.front();
            Radiants.pop();
            int curr_D = Dires.front();
            Dires.pop();

            if (curr_R < curr_D) Radiants.push(senate_size + curr_R);
            else Dires.push(senate_size + curr_D);

        }

        return Dires.empty() ? "Radiant" : "Dire";

    }
};
